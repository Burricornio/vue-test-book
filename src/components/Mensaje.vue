<template>
  <li style="margin-top: 10px" class="mensaje" @click="onClick">{{ mensaje }}</li>
</template>

<script>
export default {
  props: {
    mensaje: {
      type: String,
      required: false,
      validator: m => m.length > 1
    },
    autor: {
      type: String,
      default: 'Maria'
    }
  },
  methods: {
    onClick() {
      this.$emit('mensaje-clickado', this.mensaje)
    }
  }
  // created() {
  //   // Tengamos presente evitar la renderización anidada
  //   // cuando puede haber efectos secundarios (“side effects”),
  //   // dado que los life cycle hooks de los componentes hijos
  //   // (created, mounted) serán disparados y por lo tanto puede haber llamadas HTTP u otros efectos secundarios que no queremos que se ejecuten. Para probar lo que comentamos, puedes probar a añadir en el componente Mensaje.vue un console.log en el hook created():
  //   console.log('Creado')
  // }
}
</script>
