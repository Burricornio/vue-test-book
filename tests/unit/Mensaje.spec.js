import { shallowMount } from '@vue/test-utils'
import Mensaje from '@/components/Mensaje.vue'


describe('Mensaje.vue', () => {
  // Vamos a utilizar un función factoría para crear el componente mensaje, pasándole propiedades:
  const createWrapper = propsData => shallowMount(Mensaje, { propsData })
  let wrapper

  it('Tiene el mismo HTML', () => {
    wrapper = createWrapper({ mensaje: 'Hola'})
    expect(wrapper.element).toMatchSnapshot()
  })

  it('Tiene la propiedad mensaje', () => {
    wrapper = createWrapper({ mensaje: 'Hola'})
    expect(wrapper.props().mensaje).toBe('Hola')
  })

  it('No tiene una propiedad llamada gato', () => {
    wrapper = createWrapper({ gato: 'Hola' })
    expect(wrapper.props().gato).toBeUndefined()
  })

  it('Tiene un atributo non-prop llamado "gato"', () => {
    wrapper = createWrapper({ gato: 'Hola' })
    expect(wrapper.attributes().gato).toBe('Hola')
  })

  it('María es la autora por defecto', () => {
    wrapper = createWrapper()
    expect(wrapper.vm.autor).toBe('Maria')
  })

  it('Comprobams validaciones en la prop mensaje', () => {
    const mensaje = createWrapper().vm.$options.props.mensaje
    // Mensaje es de tipo cadena
    expect(mensaje.type).toBe(String)
    // MMensaje es obligatorio
    expect(mensaje.required).toBe(false)
    // Mensaje tiene al menos dos caracteres
    expect(mensaje.validator && mensaje.validator('a')).toBeFalsy()
    expect(mensaje.validator && mensaje.validator('aa')).toBeTruthy()
  })

  describe('Eventos', () => {

    beforeEach(() => {
      wrapper = createWrapper({ mensaje: 'Lemmy' })
    })

    it('llama a "onClick" cuando se clicka en un mensaje-item', () => {
      // stub significa en testing un objeto sustituido
      const stub = jest.fn()
      wrapper.setMethods({ onClick: stub })

      wrapper.find('.mensaje').trigger('click')
      expect(stub).toBeCalled()

      // Otra forma de ver si se ha lanzado una vez
      // wrapper.find('.mensaje').trigger('click')
      // expect(wrapper.emitted('mensaje-clickado')).toHaveLength(1)
    })

    it('emite un evento "mensaje-clickado" cuando el método onClick se invoca', () => {
      const stub = jest.fn()
      wrapper.vm.$on('mensaje-clickado', stub)
      wrapper.vm.onClick()

      expect(stub).toBeCalledWith('Lemmy')
    })
  })
})
