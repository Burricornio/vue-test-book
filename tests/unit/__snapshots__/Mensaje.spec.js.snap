// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mensaje.vue Tiene el mismo HTML 1`] = `
<li
  class="mensaje"
  style="margin-top: 10px;"
>
  Hola
</li>
`;
